presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -71,44 +71,99 @@ class PartitionAndSerializeOperator : public Operator {
 
   void addInput(RowVectorPtr input) override {
     input_ = std::move(input);
+    if (replicateNullsAndAny_) {
+      collectNullRows();
+    }
+    processingFirstBatch_ = true;
   }
 
+  /// Based on 'replicateNullsAndAny_' we will process input vector differently.
+  /// When 'replicateNullsAndAny_' is true with large numPartitions_, the output
+  /// vector size could be significantly large. To handle such large output
+  /// vector, we will yield output in batch. There are total of 3 branch of
+  /// execution: Branch 1) 'replicateNullsAndAny_' is false, partition and
+  /// serialize input elements only.
+  ///
+  /// Branch 2) 'replicateNullsAndAny_' is true and output vector
+  /// is of reasonable size compared to desired output size per batch (see:
+  /// Operator::outputBatchRows).
+  ///
+  /// Branch 3) 'replicateNullsAndAny_' is true and
+  /// output vector will be significantly large, Do this in steps:
+  ///
+  /// Step 1) Same as Branch 1.
+  ///
+  /// Step 2) Replicate null values for 'nextPartitionId_' and repeat with
+  /// nextPartitionId_ + 1
   RowVectorPtr getOutput() override {
     if (!input_) {
       return nullptr;
     }
-    collectNullRows();
 
-    const vector_size_t maxOutputSize = calculateMaxOutputSize();
+    vector_size_t numInput = input_->size();
+    velox::row::UnsafeRowFast unsafeRow(reorderInputsIfNeeded());
+    calculateSerializedRowSize(unsafeRow);
 
-    // TODO Reuse output vector.
-    auto output =
-        BaseVector::create<RowVector>(outputType_, maxOutputSize, pool());
+    std::optional<uint64_t> averageRowSize;
+    if (!cumulativeRowSizes_.empty()) {
+      averageRowSize = cumulativeRowSizes_[cumulativeRowSizes_.size() - 1] /
+          cumulativeRowSizes_.size();
+    }
+    const vector_size_t maxOutputSize = outputBatchRows(averageRowSize);
+    auto output = createOutputRowVector(maxOutputSize);
     auto partitionVector = output->childAt(0)->asFlatVector<int32_t>();
-    partitionVector->resize(maxOutputSize);
     auto dataVector = output->childAt(1)->asFlatVector<StringView>();
-    dataVector->resize(maxOutputSize);
-
-    computePartitions(*partitionVector);
-    serializeRows(*dataVector);
-
-    if (replicateNullsAndAny_ && numPartitions_ > 1) {
-      auto numOutputRows =
-          replicateRowsWithNullPartitionKeys(*dataVector, *partitionVector);
-      // If numOutputRows equal to the input row count, then
-      // 'replicateRowsWithNullPartitionKeys' did not insert any row into the
-      // output vector, thus we need to call into
-      // 'replicateAnyToEmptyPartitions'
-      if (numOutputRows == input_->size()) {
-        numOutputRows = replicateAnyToEmptyPartitions(
-            *partitionVector, *dataVector, numOutputRows);
+
+    if (processingFirstBatch_) {
```

Comment:
This method is rather long. Would it make sense to extract the body of this 'if' statement into a helper method: processFirstBatch()